core/variable_operations.cpp

```diff
@@ -294,4 +294,77 @@ Variable masks_merge_if_contained(const MasksConstProxy &masks,
   return mask_union;
 }
 
+VariableProxy nan_to_num(const VariableConstProxy &var,
+                         const VariableConstProxy &replacement,
+                         const VariableProxy &out) {
+
+  if (var.dtype() != replacement.dtype())
+    throw except::TypeError("Replacement type doesn't match type of input");
+
+  transform_in_place<pair_self_t<double, float>>(
+      out, var,
+      scipp::overloaded{
+          [&](auto &out_v, const auto &in_v) {
+            using V_OUT = std::decay_t<decltype(out_v)>;
+            using V_IN = std::decay_t<decltype(in_v)>;
+            if constexpr (is_ValueAndVariance_v<V_OUT>) {
+              if constexpr (is_ValueAndVariance_v<V_IN>) {
+                if (std::isnan(in_v.value)) {
+                  out_v.value = replacement.value<decltype(V_IN::value)>();
+                  out_v.variance =
+                      replacement.hasVariances()
+                          ? replacement.variance<decltype(V_IN::value)>()
+                          : out_v.variance = out_v.value;
+                } else {
+                  // Assign directly to output
+                  out_v.value = in_v.value;
+                  out_v.variance = in_v.variance;
+                }
+              } else {
+                if (std::isnan(in_v)) {
+                  out_v.value = replacement.value<V_IN>();
+                  out_v.variance = replacement.hasVariances()
+                                       ? replacement.variance<V_IN>()
+                                       : out_v.value;
+                } else {
+                  // Assign directly to output
+                  out_v.value = in_v;
+                  out_v.variance = in_v;
+                }
+              }
+            } else {
+              out_v = std::isnan(in_v) ? replacement.value<V_IN>() : in_v;
+            }
+          },
+          [&](units::Unit &ua, const units::Unit &ub) {
+            expect::equals(ua, ub);
+          }});
+  return out;
+}
+
+Variable nan_to_num(const VariableConstProxy &var,
+                    const VariableConstProxy &replacement) {
+  return transform<double, float>(
+      var,
+      overloaded{
+          [&](const auto &x) {
+            using V = std::decay_t<decltype(x)>;
+            if constexpr (is_ValueAndVariance_v<V>) {
+              if (std::isnan(x.value)) {
+                auto val = replacement.value<decltype(V::value)>();
+                if (replacement.hasVariances()) {
+                  // Use the specified variance in replacement
+                  return V{val, replacement.variance<decltype(V::variance)>()};
+                } else
+                  return V{val, val};
+              } else {
+                return x; // Nothing to replace
+              }
+            } else {
+              return std::isnan(x) ? replacement.value<V>() : x;
+            }
+          },
+          [](const units::Unit &u) { return u; }});
+}
```

Comment:
Have you considered using `transform` with 2 arguments? Then your lambda is just

```cpp
[](const auto &x, const auto &replacement) { return isnan(x) ? replacement : x; }
```
?

Reply:
I've simplified, unfortunately your suggestion would not work as is because in the absence of variances, you get a different type (applies for both arguments). So you still need to swtich between nan handling with/without variances.